pkg/controllers/status/common.go

```diff
@@ -69,13 +72,28 @@ func updateResourceStatus(
 		klog.Errorf("Failed to aggregate status for resource(%s/%s/%s, Error: %v", gvr, resource.GetNamespace(), resource.GetName(), err)
 		return err
 	}
-	if reflect.DeepEqual(resource, newObj) {
+
+	oldStatus, _, _ := unstructured.NestedFieldNoCopy(resource.Object, "status")
+	newStatus, _, _ := unstructured.NestedFieldNoCopy(newObj.Object, "status")
+	if reflect.DeepEqual(oldStatus, newStatus) {
 		klog.V(3).Infof("Ignore update resource(%s/%s/%s) status as up to date.", gvr, resource.GetNamespace(), resource.GetName())
 		return nil
 	}
 
-	if _, err = dynamicClient.Resource(gvr).Namespace(resource.GetNamespace()).UpdateStatus(context.TODO(), newObj, metav1.UpdateOptions{}); err != nil {
-		klog.Errorf("Failed to update resource(%s/%s/%s), Error: %v", gvr, resource.GetNamespace(), resource.GetName(), err)
+	var statusPatch []byte
+	if newStatus != nil {
+		j, err := json.Marshal(newStatus)
+		if err != nil {
+			klog.Errorf("Failed to marshal status for resource(%s/%s/%s, Error: %v", gvr, resource.GetNamespace(), resource.GetName(), err)
+			return err
+		}
+		statusPatch = []byte(fmt.Sprintf(`[{"op": "replace", "path": "/status", "value": %s}]`, j))
+	} else {
+		statusPatch = []byte(`[{"op": "remove", "path": "/status"}]`)
+	}
```

Comment:
Is it possible to consider using `helper.GenMergePatch`?

Reply:
@XiShanYongYe-Chang The reason I used JSONPatchType instead of MergePatchType is that I am worried that some fields may not be deleted when patch status, because we do not understand and cannot modify the definition of resource templates and the logic of related controllers. We even provide a webhook so that users can customize the status.

If you think jsonpatch is acceptable, I can move this part of the construction logic to helper/patch.go (I didn't notice this file existed before); or if you think merge is still a better choice, Then I will update this part.


Obviously I should reply here...😆

Reply:
:) haha

Reply:
I have done some research and conducted some tests, and it seems that MergePatch can completely replace JSONPatch. Therefore, I choose to use MergePatch here. We can monitor if anyone reports any issues in the future.

Reply:
done.

Reply:
Never mind. Full discussion and testing gave us a much better result 👍

Reply:
Updated. kindly ping @XiShanYongYe-Chang to review again.

Reply:
Thank you for your explanation, I probably understand what you mean, can you help write an example to describe it, so that it is easier to follow?

Reply:
A simple example could be like this: [playground](https://go.dev/play/p/w0GjAogpevo).
If everything works on time but the informer holding an outdated resource template (currentA), using mergePatch may create an inappropriate mergePatchBytes for actual template object (currentB).